cpp/src/unary/cast_ops.cu

```diff
@@ -0,0 +1,123 @@
+  CUDA_DEVICE_CALLABLE R operator()(T const element) {
+    return static_cast<R>(element);
+  }
+  template <typename T = _T,
+            typename R = _R,
+            typename std::enable_if_t<cudf::is_numeric<T>() &&
+                                      cudf::is_timestamp<R>()>* = nullptr>
+  CUDA_DEVICE_CALLABLE R operator()(T const element) {
+    return static_cast<R>(static_cast<typename R::rep>(element));
+  }
+  template <typename T = _T,
+            typename R = _R,
+            typename std::enable_if_t<cudf::is_timestamp<T>() &&
+                                      cudf::is_numeric<R>()>* = nullptr>
+  CUDA_DEVICE_CALLABLE R operator()(T const element) {
+    return static_cast<R>(element.time_since_epoch().count());
+  }
+};
+
+template <typename T>
+struct dispatch_unary_cast_to {
+  column_view input;
+  mutable_column_view output;
+
+  dispatch_unary_cast_to(column_view inp, mutable_column_view out)
+      : input(inp), output(out) {}
+
+  template <typename R,
+            typename std::enable_if_t<cudf::is_numeric<R>() ||
+                                      cudf::is_timestamp<R>()>* = nullptr>
+  void operator()(cudaStream_t stream) {
+    thrust::transform(rmm::exec_policy(stream)->on(stream), input.begin<T>(),
+                      input.end<T>(), output.begin<R>(), unary_cast<T, R>{});
+  }
+
+  template <typename R,
+            typename std::enable_if_t<!cudf::is_numeric<R>() &&
+                                      !cudf::is_timestamp<R>()>* = nullptr>
+  void operator()(cudaStream_t stream) {
+    CUDF_FAIL("Column type must be numeric");
+  }
+};
+
+struct dispatch_unary_cast_from {
+  column_view input;
+  mutable_column_view output;
+
+  dispatch_unary_cast_from(column_view inp, mutable_column_view out)
+      : input(inp), output(out) {}
+
+  template <typename T,
+            typename std::enable_if_t<cudf::is_numeric<T>() ||
+                                      cudf::is_timestamp<T>()>* = nullptr>
+  void operator()(cudaStream_t stream) {
+    experimental::type_dispatcher(
+        output.type(), dispatch_unary_cast_to<T>{input, output}, stream);
+  }
+
+  template <typename T,
+            typename std::enable_if_t<!cudf::is_timestamp<T>() &&
+                                      !cudf::is_numeric<T>()>* = nullptr>
+  void operator()(cudaStream_t stream) {
+    CUDF_FAIL("Column type must be numeric");
+  }
+};
+}  // namespace detail
+
+std::unique_ptr<column> cast(column_view const& input,
+                             data_type type,
+                             rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS(is_fixed_width(type), "Unary cast type must be fixed-width.");
+
+  auto size = input.size();
+  auto null_mask = copy_bitmask(input, 0, mr);
+  auto output = std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), 0, mr},
+      null_mask, input.null_count(), std::vector<std::unique_ptr<column>>{});
```

Comment:
Careful, this was deep copying the device buffer.
```suggestion
      std::move(null_mask), input.null_count(), std::vector<std::unique_ptr<column>>{});
```

Reply:
Don't we want to deep-copy the device buffer? What happens if the input column is still valid and in use elsewhere?